framework/web/UrlManager.php

```diff
@@ -309,18 +310,38 @@ public function createUrl($params)
         $baseUrl = $this->showScriptName || !$this->enablePrettyUrl ? $this->getScriptUrl() : $this->getBaseUrl();
 
         if ($this->enablePrettyUrl) {
+            $cacheKey = $route . '?' . implode('&', array_keys($params));
+
             /* @var $rule UrlRule */
-            foreach ($this->rules as $rule) {
-                if (($url = $rule->createUrl($this, $route, $params)) !== false) {
-                    if (strpos($url, '://') !== false) {
-                        if ($baseUrl !== '' && ($pos = strpos($url, '/', 8)) !== false) {
-                            return substr($url, 0, $pos) . $baseUrl . substr($url, $pos);
-                        } else {
-                            return $url . $baseUrl . $anchor;
-                        }
+            $url = false;
+            if (isset($this->_createUrlRuleMatches[$cacheKey])) {
+                foreach ($this->_createUrlRuleMatches[$cacheKey] as $rule) {
+                    if (($url = $rule->createUrl($this, $route, $params)) !== false) {
+                        break;
+                    }
+                }
+            } else {
+                $this->_createUrlRuleMatches[$cacheKey] = [];
+            }
+
+            if ($url === false) {
+                foreach ($this->rules as $rule) {
+                    if (($url = $rule->createUrl($this, $route, $params)) !== false) {
+                        $this->_createUrlRuleMatches[$cacheKey][] = $rule;
+                        break;
+                    }
+                }
+            }
+
+            if ($url !== false) {
+                if (strpos($url, '://') !== false) {
+                    if ($baseUrl !== '' && ($pos = strpos($url, '/', 8)) !== false) {
+                        return substr($url, 0, $pos) . $baseUrl . substr($url, $pos);
                     } else {
-                        return "$baseUrl/{$url}{$anchor}";
+                        return $url . $baseUrl . $anchor;
```

Comment:
This is not a change. GIT just compose a 'smart' diff, so it is look that way.
This is actually a former line 319 (see above). It just have been moved.
Original line now is line 344 - it is intact.